### PR TITLE
GH-27: Support to retrieve udids from network

### PR DESCRIPTION
**Pull request by didix21**
- Modify behaviour of idevice_id:

  - Now if options usblist and networklist are set, idevice_id is going
    to return the list of iPhones / iPad udids found pluged on the usb
    or finded in the network.

  - Replace how errors were reported from native libimobiledevice. Now
    return number codes which are handled in javascript with the
    following error types: WrongParameterTypeError,
    CannotRetrieveDeviceListError, CannotMallocMemoryError,
    CannotReallocMemoryError.

- Move id_worked to lib/worker path.

- Move javascript business logic of idevce_id from index.js to
  lib/idevice_id.js.

- Remove MaybeLocal in ideviceinfo.

- Test idevice_id.

- Add mocha, chain and sinon packages.

### CODE REVIEW

File: lib/workers/id_worker.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const lib = require(`${__dirname}/../../build/Release/imobiledevice.node`)
+
+process.on('message', options => {
+	lib.idevice_id(options, (err, data) => {
+		if (err) process.send({err: err})
+		else process.send({data: data})
```
Comment (formonkey) [inline lib/workers/id_worker.js:6]: process.send((err) ? { err } : { data });

File: test/idevice_id.spec.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
+
+const { id } = require('../index');
+const cp = require('child_process');
+var assert = require('assert');
```
Comment (formonkey) [inline test/idevice_id.spec.js:4]: Change var to let or const

File: test/idevice_id.spec.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+const { expect } = require('chai');
+const { WrongParameterTypeError, CannotRetrieveDeviceListError, CannotMallocMemoryError } = require('../index');
+const { stub } = require('sinon');
+const {native_id_errors, CannotReallocMemoryError} = require('../lib/idevice_id');
+
+
+describe('Test idevice_id', () => {
+    
+    const stubChildOn = (error, devicelist) => { return (text, callback) => callback({err: error, data: devicelist}) };
```
Comment (formonkey) [inline test/idevice_id.spec.js:13]: Not need use return

`const stubChildOn = (error, devicelist) => (text, callback) => callback({err: error, data: devicelist});`

File: lib/idevice_id.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                    break;
+                case native_id_errors.ID_E_CANNOT_RETRIEVE_DEVICE_LIST:
+                    callback(new CannotRetrieveDeviceListError(), null);
+                    break;
+                case native_id_errors.ID_E_CANNOT_MALLOC_MEMORY:
+                    callback(new CannotMallocMemoryError(), null);
+                    break;
+                case native_id_errors.ID_E_CANNOT_REALLOC_MEMORY:
+                    callback(new CannotReallocMemoryError(), null);
+                    break;
```
Comment (formonkey) [inline lib/idevice_id.js:62]: Don't need a default error?

File: lib/idevice_id.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        this.code = native_id_errors.ID_E_CANNOT_REALLOC_MEMORY;
+    }
+}
+
+exports.idevice_id = function(options, callback) {
+	if (typeof options === "function") {
+		callback = options;
+		options = null;
+        options = options || {};
+	}
```
Comment (formonkey) [inline lib/idevice_id.js:40]: `
if (typeof options === 'function') {
    callback = options;
    options = {};
}
`

File: lib/idevice_id.js (hunk 4 above)
Comment (didix21) [reply lib/idevice_id.js:62]: good observation! I'm going to take into account thanks

File: lib/idevice_id.js (hunk 5 above)
Comment (didix21) [reply lib/idevice_id.js:40]: Maybe I will remove this, I don't really sure if it has a lot of sense.

**Review by formonkey — APPROVED** [review]

**Pull request merged**